### DISCUSSION TITLE
fpm support for GraalVM

### DISCUSSION THREAD

**Discussion by schrepfler** [fortran-lang/fpm#989] [General]
Hi, I've recently started to take a look at the GraalVM Java distribution and trying to understand if it's possible to link and use a [fortran library that I've found on github](https://github.com/jacobwilliams/PowellOpt).
I noticed the project was managed by fpm and was pleasantly surprised how easy it was to setup on a MacOS ARM laptop and how it managed to download the toolchain and build the project and run the test. Kudos, that's a great DX, wasn't expecting from a 66 years old language! 
I noticed the build tool outputted 3 directories in the build directory (can maybe someone explain in layman terms what do .a, .a.log, and .mod files represent in Fortran idiom? 

```
tree build
build
├── cache.toml
├── dependencies
├── gfortran_50F62D7499E64B65
│   └── PowellOpt
│       ├── libPowellOpt.a
│       └── libPowellOpt.a.log
├── gfortran_AC789FC96BC46CEE
│   ├── PowellOpt
│   │   ├── src_bobyqa.f90.o
│   │   ├── src_bobyqa.f90.o.digest
│   │   ├── src_bobyqa.f90.o.log
│   │   ├── src_cobyla.f90.o
│   │   ├── src_cobyla.f90.o.digest
│   │   ├── src_cobyla.f90.o.log
│   │   ├── src_kind_module.f90.o
│   │   ├── src_kind_module.f90.o.digest
│   │   ├── src_kind_module.f90.o.log
│   │   ├── src_lincoa.f90.o
│   │   ├── src_lincoa.f90.o.digest
│   │   ├── src_lincoa.f90.o.log
│   │   ├── src_newuoa.f90.o
│   │   ├── src_newuoa.f90.o.digest
│   │   ├── src_newuoa.f90.o.log
│   │   ├── src_powellopt.f90.o
│   │   ├── src_powellopt.f90.o.digest
│   │   ├── src_powellopt.f90.o.log
│   │   ├── src_uobyqa.f90.o
│   │   ├── src_uobyqa.f90.o.digest
│   │   ├── src_uobyqa.f90.o.log
│   │   ├── test_tests.f90.o
│   │   ├── test_tests.f90.o.digest
│   │   └── test_tests.f90.o.log
│   ├── bobyqa_module.mod
│   ├── cobyla_module.mod
│   ├── kind_module.mod
│   ├── lincoa_module.mod
│   ├── newuoa_module.mod
│   ├── powellopt.mod
│   └── uobyqa_module.mod
└── gfortran_E167FD2A985B468F
    └── test
        ├── tests
        └── tests.log
```

My question is following, GraalVM ships with a [LLVM Runtime](https://www.graalvm.org/latest/reference-manual/llvm/) that contains a tool `lli` (not sure if that's a standard LLVM command) which looks like can compile C/C++, Fortran and other languages  like a normal llvm toolchain and make it interoperable with the GraalVM and thus in [polyglot Java](https://www.graalvm.org/latest/reference-manual/llvm/Interoperability/). Is there a way to somehow tell fpm to use this toolchain to produce GraalVM output or how easy would it be to extend fpm with build support for this (different build targets)? Is there anyone in the fpm which is working on this or has interest in this?

**Comment by schrepfler** [comment]
I did notice in the documentation for the R interop that it seems they too use gfortran and don't use llvm bitcode but use fortran in GraalVM R runtime native mode vs LLVM mode.

**Reply by ivan-pi** [reply]
You may have a way of calling compiled Fortran as native code if that's the case.

**Comment by ivan-pi** [comment]
`.a` are static libraries (archives) packed with object files in ELF format. `.mod` are Fortran module files that are specific to each compiler. They are needed at compilation time. `.log` and `.digest` are meta-data used by fpm. 

For Powell's algorithms, have a look at PRIMA (https://github.com/libprima/prima), which is a refactored and very heavily tested implementation.

**Reply by schrepfler** [reply]
Thanks Ivan!
